DEMO-1 quarto notebook (run workflow from pecan.xml) (#3531)
PecanProject/pecan: documentation/quarto/01_Demo_Basic_Run/run_pecan.qmd

---

> title: "PEcAn Workflow"

Reviewer: As [discussed above](https://github.com/PecanProject/pecan/pull/3531/files/6a8ae417d4922de34b1b216e2b1791c0991dcae3#r2183396745), I don't think it is useful to have a `quarto` subdirectory. I prefer `documentation/01_demo_basic_run.qmd` but `documentation/01_demo_basic_run/run_pecan.qmd` also makes sense.

Author: moved to this directory `pecan/documentation/tutorials/Demo_1_Basic_Run/run_pecan.qmd`

author: "PEcAn"
format: pdf

> # Introduction {#introduction}

Reviewer: this is a well written introduction! Clear and easy to follow.

Welcome to this PEcAn workflow notebook! This notebook will guide you through running an ecosystem model using PEcAn's programmatic interface. 

## What is PEcAn?

PEcAn (Predictive Ecosystem Analyzer) is a scientific workflow system designed to make ecosystem modeling more transparent, repeatable, and accessible. It helps researchers:

- Run ecosystem models with standardized inputs and outputs
- Perform uncertainty analysis on model parameters
- Compare model predictions with observations
- Share and reproduce scientific workflows

## What This Notebook Does

This notebook demonstrates how to:
1. Set up and configure a PEcAn workflow
2. Run an ecosystem model simulation
3. Analyze and visualize the results

## Prerequisites

Before running this notebook, make sure you have:

> - The PEcAn R packages installed

Reviewer: How do I do this if they are not already installed? Link to https://github.com/pecanProject/pecan?tab=readme-ov-file#installation

Author: This point is not needed and I will remove it. When someone runs a PEcAn setup locally, it usually runs inside their Docker stack, so there is no need to specify this point separately.

> - A valid `pecan.xml` configuration file (or use the example provided)

Reviewer: Link to example provided

## How to Use This Notebook

1. Each section is clearly marked with a heading
2. Code chunks are provided with explanations
3. You can run the code chunks sequentially

> 4. Feel free to modify parameters to suit your needs

Reviewer: ```suggestion
4. Once you have successfully run the demo, you can modify parameters to configure new runs and analyses. 
```

Let's get started!

#### Objective

This demo illustrates how to run a basic PEcAn workflow using an R-based Quarto notebook. It will cover loading settings, writing model configuration files, and running model simulations. This approach provides a programmatic alternative to the web-based PEcAn interface for executing ecosystem models.

# Load PEcAn packages

First, we need to load the PEcAn R packages. These packages provide all the functions we'll use to run the workflow.

```{r libraries}
# Load the PEcAn.all package, which includes all necessary PEcAn functionality
library("PEcAn.all")
```

# Load PEcAn settings files

This section handles the crucial step of loading the PEcAn settings. 

> -   If you have a `pecan.xml` file precofigured, place it in your working directory and copy its file path. Assign this path to the **settings_path** variable.

Reviewer 2: fix typos

> -   If you have a `pecan.xml` file precofigured, place it in your working directory and copy its file path. Assign this path to the **settings_path** variable.

Reviewer: Where would a new user get a `pecan.xml` file that they can be confident will work on their computer?

Author: One useful step could be to direct users to refer to a settings file from the `tests/` directory. However, many of the files in that directory contain outdated configurations, such as old database settings(though database is not required for this workflow), which may not work correctly. For now, it’s better to redirect users to the example settings file instead.

-   Alternatively, for a sample or initial run, you can use the example `pecan.xml` file available at: `pecan/documentation/quarto/01_Demo_Basic_Run`.
 
```{r load-settings}
# Specify the path to your pecan.xml file
# Replace this with your actual pecan.xml file path

> settings_path <- "~/pecan.xml" 

Reviewer: Ideally, this workflow will work 'out of the box' on anyone's MacOS or Linux computer. So examples should point to the actual file, e.g. 

```
settings_path <- "documentation/quarto/01_Demo_Basic_Run/pecan.xml"
```

Note that the `here` package can help ensure that the relative path is correct, whether running it interactively or compiling:

```
settings_path <- here::here("documentation/quarto/01_Demo_Basic_Run/pecan.xml")
```

```

# Prepare and Validate Settings

After specifying the path to your `pecan.xml` file, the next step involves reading and preparing these settings. PEcAn provides robust utilities to process and validate your configurations before any execution begins. 

*   `PEcAn.settings::read.settings(settings_path)`: This function parses the `pecan.xml` file, converting its contents into an R list object that PEcAn can work with. This step ensures that the XML structure is correctly interpreted.
*   `PEcAn.settings::prepare.settings(settings)`: After reading, the settings are passed to this function for further preparation and validation. This involves checking for missing required fields, setting up default values where necessary, and ensuring that all paths and configurations are consistent with PEcAn's operational requirements. This step is crucial for preventing errors during subsequent workflow stages.

```{r read-prepare-settings}
# Read the settings from the pecan.xml file
settings <- PEcAn.settings::read.settings(settings_path)

# Prepare and validate the settings
settings <- PEcAn.settings::prepare.settings(settings)
```

# Exploring the settings objects

> Once the settings have been read and prepared, it is often useful to inspect the structure of the `settings` object. This object is a comprehensive R list that contains all the parameters and configurations for your PEcAn workflow, derived directly from the `pecan.xml` file. 

Reviewer: ```suggestion
Once the settings have been read and prepared, it is often useful to inspect the structure of the `settings` object. This object is the R list that we just created from the `pecan.xml` file; it contains parameters and configurations for the PEcAn workflow. 
```

> *   `str(settings)`: This function provides a concise, human-readable summary of the structure of the `settings` object. It displays the internal structure of an R object, showing its type, length, and the first few elements for each component. This helps in understanding how your XML configurations are represented in R and can be invaluable for debugging or verifying settings during the workflow.

Reviewer: because we aren't teaching users about `str`, I think this could be simplified
```suggestion
*   `str(settings)` displays the internal structure of the `settings` object. This shows how the settings are represented in R and is useful for debugging and verifying settings.
```

```{r explore-settings}
# Display the structure of the settings object
# This helps you understand what configurations are available
str(settings)
```
# Run Model Simulations and Fetch Results

This section executes the actual model simulations and retrieves the results. The process is managed by PEcAn's workflow system, which handles the execution of your chosen ecosystem model.

> *   `start_model_runs(settings)`: This function initiates the model runs based on your configuration. It manages the execution of your chosen ecosystem model, using the configuration files generated in the previous step.

Reviewer 2: Having a call to `settings <- PEcAn.workflow::runModule.run.write.configs(settings)` isn't optional -- you need to add that step before calling start_model_runs.  Indeed, it's probably the least optional function in all of PEcAn.

```{r run-model}
# Start the model simulations
PEcAn.workflow::start_model_runs(settings)
```

> 

Reviewer 2: seems like there should be some text here before the this block explaining what the goal of the code chunk is

```{r get-plot-vars}

> runid<-as.character(read.table(paste(settings$outdir, "/run/","runs.txt", sep=""))[1,1]) # Note: if you are using an xml from a run with multiple ensembles this line will provide only the first run id

Reviewer: See the tidyverse style guide - in particular, there should be more spaces here, and consider using the `styler` package. (we haven't done / don't do that consistently or on existing code, but it is relatively easy to use and is a good idea for new contributions) https://style.tidyverse.org/

Author: thanks for this.helps a lot in indentation.

outdir<- paste(settings$outdir,"/out/",runid,sep= "")
start.year<-as.numeric(lubridate::year(settings$run$start.date))
end.year<-as.numeric(lubridate::year(settings$run$end.date))

# Load PEcAn output using read.output()
model_output <- PEcAn.utils::read.output(
  runid,
  outdir,
  start.year,
  end.year,
  variables = NULL,  
  dataframe = TRUE, 
  verbose = FALSE  
)

# Store variable names for later use
available_vars <- names(model_output)[!names(model_output) %in% c("posix", "time_bounds")]
```

```{r show-variables, echo=FALSE}

> # Display available variables in a clean format

Reviewer: If it helps, I've documented these here (feel free to copy-paste) https://github.com/ccmmf/downscaling/blob/3e4d0680a165d3ede7b91c8ad0709dd151ead4f2/docs/workflow_documentation.md?plain=1#L184. 

But it would be even better to generate such a table dynamically (to ensure the variables and their defenitions up to date).

And its not clear that you need the list in both the r chunk comments and displayed in the compiled document. If you want them in comments for users to see in the unrendered version, you can use `<!-- markdown comments -->` outside the code chunk.

```r
standard_vars |> 
  dplyr::filter(Variable.Name %in% available_vars) |> 
  dplyr::select(Variable.Name, Units) |> #optionally include Description 
  knitr::kable()
```

cat("Available variables in model output:\n")
cat(paste(available_vars, collapse = "\n"))
```

```{r plot-timeseries}
# You can see all available variables in the output from the previous block above
# They are also listed here for reference:

# Carbon fluxes:
# - GPP, NPP, NEE, TotalResp, AutoResp, HeteroResp, SoilResp
# Carbon pools:
# - AbvGrndWood, TotLivBiom, TotSoilCarb, leaf_carbon_content
# - litter_carbon_content, fine_root_carbon_content, coarse_root_carbon_content
# - GWBI, AGB
# Water variables:
# - Qle, Transp, SoilMoist, SoilMoistFrac, SWE
# - litter_mass_content_of_water
# Other:
# - LAI

# Example 1: Carbon fluxes
plot(model_output$posix, model_output$GPP, 
     type = 'l',
     col = 'green',
     xlab = 'Date',
     ylab = 'Carbon Flux (kg C m-2 s-1)',
     main = 'Carbon Fluxes Over Time')

# Add NPP line
lines(model_output$posix, model_output$NPP, 
      col = 'blue')

# Add legend
legend('topright', 
       legend = c('GPP', 'NPP'),
       col = c('green', 'blue'),
       lty = 1)

# Example 2: Carbon pools

> # To plot carbon pools instead, uncomment and modify these lines:

Reviewer 2: Rather than commenting out lines and having the user uncomment them, I'd just put them in separate code chunks

Reviewer: Agreed - it would be helpful for these to appearh in the rendered document

start.year<-as.numeric(lubridate::year(settings$run$start.date))
end.year<-as.numeric(lubridate::year(settings$run$end.date))

# Load PEcAn output using read.output()
model_output <- PEcAn.utils::read.output(
  runid,
  outdir,
  start.year,
  end.year,
  variables = NULL,  
  dataframe = TRUE, 
  verbose = FALSE  
)

# Store variable names for later use
available_vars <- names(model_output)[!names(model_output) %in% c("posix", "time_bounds")]
```

```{r show-variables, echo=FALSE}
# Display available variables in a clean format
cat("Available variables in model output:\n")
cat(paste(available_vars, collapse = "\n"))
```

```{r plot-timeseries}
# You can see all available variables in the output from the previous block above
# They are also listed here for reference:

# Carbon fluxes:
# - GPP, NPP, NEE, TotalResp, AutoResp, HeteroResp, SoilResp
# Carbon pools:
# - AbvGrndWood, TotLivBiom, TotSoilCarb, leaf_carbon_content
# - litter_carbon_content, fine_root_carbon_content, coarse_root_carbon_content
# - GWBI, AGB
# Water variables:
# - Qle, Transp, SoilMoist, SoilMoistFrac, SWE
# - litter_mass_content_of_water
# Other:
# - LAI

# Example 1: Carbon fluxes

> plot(model_output$posix, model_output$GPP, 

Author: I've added additional examples here, as I thought they might be helpful for novice users.

Reviewer: That is great, and helpful!
